include/python_partition_interface.h

```diff
@@ -28,10 +28,9 @@ MutableVertexPartition* create_partition(Graph* graph, char* method, vector<size
 MutableVertexPartition* create_partition_from_py(PyObject* py_obj_graph, char* method, PyObject* py_initial_membership, PyObject* py_weights, PyObject* py_node_sizes, double resolution_parameter);
 
 Graph* create_graph_from_py(PyObject* py_obj_graph);
```

Comment:
You are right, the version w/o the node_sizes became redundant. I've removed it.

As for what happens if a node is inside a too-large partition....

TL;DR: I think the current pull-request stands on its own, and I'd be more than happy to discuss more ambitious follow-up work.

The (too long) answer is that I did think about it and my reasoning was as follows:

I viewed `max_comm_size` as an operational directive saying "do not add more nodes to _create_ too-large communities". This is a natural minor tweak to the algorithm, and if you apply this constraint to _all_ the phases, you are guaranteed to have not-too-large communities. Call this the "weak form" of the constraint. It seems useful on its own and can be used as a basic step for a more general approach.

I did not view `max_comm_size` as a more general directive of "ensure all communities are up to this size no matter what". I toyed with the idea, together with the "natural" accompanying `min_comm_size` to force the community sizes to be in some desired range. Call this the "strong form" of the constraint.

The "strong form" is basically what I need - metacells can't be either too small or too large. Right now I do this by using the "weak form" as a basic step.

I decided not to directly add "strong form" logic to the basic algorithm because this would no longer be a minor natural tweak to the algorithm. The current algorithm is a nice and simple bottom-up approach (ok, maybe not _that_ simple). But the optimal algorithm for establishing size constraints (especially given an initial partition) might no longer be a "simple" tweak of this approach.

I'm not certain how to _best_ split too-large communities, or combine too-small ones. I suspect this doesn't necessarily map to the bottom-up phased approach. The result might end up a "Leiden++" algorithm, which would be equivalent to the Leiden algorithm if the constraints would not be used.

Given that my need for tight control over the community size isn't a common one, I was not comfortable pushing such a major change into a general-purpose package - especially as the implementation would be rather "experimental".

If you are interested in pursuing this direction, I'd be happy to cooperate in doing so:

The "naive" method would be to force the move of a node to reduce the size of a too-large community, and likewise, force the move of nodes to increase the size of a too-small community.

The problem is that these two constraints conflict with each other, so there needs to be some balanced way to avoid endlessly splitting a too-large community into too-small ones, and merging these back into a too-large communities. This can be done by having some explicit global "penalty" function for violating the constraints, and being at least somewhat "greedy"
in minimizing it when making the forced moves.

From a bit of playing around with this, however, it seems that a purely greedy approach would not work well (which isn't very surprising). A better approach might be having an explicit "improve constraints" phase that used more global operations (e.g., "split too large a community into two").

BTW, my first approach was to use the unmodified Leiden algorithm, and recursively apply it to only "too-large" communities to split them into smaller parts. Then apply the unmodified algorithm to a graph where each node is a too-small community, to merge them into larger communities, going back and forth as long as I improved the overall result.

This worked, mostly, but being able to avoid the "split too large communities" step by tweaking the basic algorithm to not generate them in the 1st place turned out to be a significant improvement (both in CPU time and quality of the result). So, what I do now is use `max_comm_size` to restrict the community size. "All" I need to do next is to do the "merge too-small communities" phase on the result. This is much simpler and seems to work well.

It seems reasonable that a more general algorithm that deals with both sides of the constraints would do even better. I did make an attempt along these lines but it didn't work too well (probably due to being too "greedy"). This is still something I might want to pursue.

In the meanwhile, I'd like to have the simple `max_comm_size` tweak in the `ledenalg` package, so I can publish my `metacells` package that depends on it, so users can start playing with it.

Sorry for the rambling answer...